crates/typst/src/math/equation.rs

```diff
@@ -223,7 +244,74 @@ impl Packed<EquationElem> {
         } else {
             vec![MathParItem::Frame(run.into_fragment(&ctx, styles).into_frame())]
         };
+        let mut pos_and_sizes: Vec<(Point, Size)> = Vec::new();
+        let mut x = Abs::zero();
 
+        // helper function to determine the vertical offset for a frame from MathParItem
+        let get_vertical_shift_fn = |frame: &Frame| {
+            let font_size = scaled_font_size(&ctx, styles);
+            let slack = ParElem::leading_in(styles) * 0.7;
+            let top_edge = TextElem::top_edge_in(styles).resolve(font_size, &font, None);
+            let ascent = top_edge.max(frame.ascent() - slack);
+            ascent - frame.baseline()
+        };
+
+        for (idx, item) in items.iter().enumerate() {
+            match item {
+                MathParItem::Frame(frame) => {
+                    // determine the coordinates of the frame in the MathParItem Array
+                    let y = get_vertical_shift_fn(frame);
+                    let pos = Point::new(x, y);
+                    let size = Size::new(frame.width().abs(), frame.height().abs());
+                    pos_and_sizes.push((pos, size));
+                    x += frame.width();
+                }
+                MathParItem::Space(space_width) => {
+                    // A MarhParItem can also be space (consumes a width), in the latter
+                    // case, we need to update the running x-coordinate
+                    // also note that we compute starting from the first non-space MathParItem
+                    if idx != 0 {
+                        x += *space_width;
+                    }
+                }
+            }
+        }
+        // computing the origin position and the size of the bounding box of the entire MathParItem
+        // Array.
+        let (pos, size) = compute_bounding_box(&pos_and_sizes);
+        let decos = TextElem::deco_in(styles);
+        let pos_and_frames = decorate_frame(&decos, pos, size);
+        let (background_pos_and_frames, foreground_pos_and_frames): (Vec<_>, Vec<_>) =
+            pos_and_frames.into_iter().partition(|&(b, _, _)| b);
+
+        let mut first_frame_index: Option<usize> = None;
+        let mut last_frame_index: Option<usize> = None;
+        for (index, item) in items.iter().enumerate() {
+            let MathParItem::Frame(_) = item else { continue };
+            if first_frame_index.is_none() {
+                first_frame_index = Some(index);
+            }
+            last_frame_index = Some(index);
+        }
+        if let Some(index) = first_frame_index {
```

Comment:
I don't really understand what's going on here. Why are we only background-decorating the first and foreground-decorating the last `MathParItem::Frame`?

Reply:
Ah, wait, are you creating an overlarge decoration for all the items, but applying it to the first one? That seems problematic when the equation breaks over multiple lines. We should probably rather create individual decorations.

Reply:
Updated. Now each frame is decorated on its own.

Special case will be any empty space between two frames, in that case we need to fill the void and decorate the space as well. For such space in between frames, we use the last frame and append the decorations to it.

Reply:
The bug is that there is a space at the end of the first line in the picture above. I think the spacing should be "weak" and collapse at the end of the line, just like normal text spaces.

Reply:
> The bug is that there is a space at the end of the first line in the picture above. I think the spacing should be "weak" and collapse at the end of the line, just like normal text spaces.

Sorry for the late update. While I have not figured out a good solution, I added a predecessor PR  that eats white space introduced by the math equation in #4087

Reply:
It's really quite tricky. Ideally, we'd also not couple math and paragraph layout too tightly. 

Maybe we should apply the highlighting completely in paragraph layout rather than math, on a per-line base? This could also fix https://github.com/typst/typst/issues/1716.